host/sgx/sgxload.c

```diff
@@ -545,24 +546,37 @@ oe_result_t oe_sgx_create_enclave(
 #elif defined(_WIN32)
 
         /* Ask OS to create the enclave */
-        DWORD enclave_error;
-        void* base = CreateEnclave(
-            GetCurrentProcess(),
-            NULL, /* Let OS choose the enclave base address */
-            secs->size,
-            secs->size,
-            ENCLAVE_TYPE_SGX,
-            (const void*)secs,
-            sizeof(ENCLAVE_CREATE_INFO_SGX),
-            &enclave_error);
+        DWORD _einit_retry_count = 0;
+        while (_einit_retry_count < MAX_EINIT_RETRY_COUNT)
+        {
+            DWORD enclave_error = 0;
+            void* base = CreateEnclave(
+                GetCurrentProcess(),
+                NULL, /* Let OS choose the enclave base address */
+                secs->size,
+                secs->size,
+                ENCLAVE_TYPE_SGX,
+                (const void*)secs,
+                sizeof(ENCLAVE_CREATE_INFO_SGX),
+                &enclave_error);
+
+            if (!base)
+            {
+                if (enclave_error == SGX_UNMASKED_EVENT)
```

Comment:
We are not going to see an error message in the worst case, we failed to create the enclave with after retrying MAX_EINIT_RETRY_COUNT times.

Reply:
Good point. Will incorporate

Reply:
Your fix still doesn't address the comment but instead spews an error message on every retry, which is not meaningful. I'm not sure why this loop needs to be so complicated even.

```c
void* base = NULL;
uint32_t enclave_error;

for (DWORD i = 0; i < MAX_EINIT_RETRY_COUNT; i++)
    base = enclave_create(
        NULL, /* Let OS choose the enclave base address */
        secs->size,
        secs->size,
        ENCLAVE_TYPE_SGX1,
        (const void*)secs,
        sizeof(sgx_secs_t),
        &enclave_error);
        
	if (enclave_error != SGX_UNMASKED_EVENT)
	    break;
}

if (!base)
{
    OE_RAISE_MSG(
        OE_PLATFORM_ERROR,
	"CreateEnclave failed (err=%#x)",
        enclave_error);
}

secs->base = (uint64_t)base;
```

Reply:
> Your fix still doesn't address the comment but instead spews an error message on every retry, which is not meaningful. I'm not sure why this loop needs to be so complicated even.
 
The error message is printed only when the retry max count is reached. Otherwise, it continues to the while loop:

                    if (einit_retry_count == MAX_EINIT_RETRY_COUNT)
                    {
                        OE_RAISE_MSG(
                            OE_PLATFORM_ERROR,
                            "CreateEnclave failed with SGX_UNMASKED_EVENT "
                            "after %d retries",
                            MAX_EINIT_RETRY_COUNT);
                    }
                    else
                        continue;
             }
OE_RAISE_MSG not only prints the error message but will jump to done and exit the routine.

I prefer for loops instead of while too.

Reply:
> Your fix still doesn't address the comment but instead spews an error message on every retry, which is not meaningful. I'm not sure why this loop needs to be so complicated even.
> 
> ```c
> void* base = NULL;
> uint32_t enclave_error;
> 
> for (DWORD i = 0; i < MAX_EINIT_RETRY_COUNT; i++)
> {
>     base = enclave_create(
>         NULL, /* Let OS choose the enclave base address */
>         secs->size,
>         secs->size,
>         ENCLAVE_TYPE_SGX1,
>         (const void*)secs,
>         sizeof(sgx_secs_t),
>         &enclave_error);
>         
>     if (enclave_error != SGX_UNMASKED_EVENT)
>         break;
> }
> 
> if (!base)
> {
>     OE_RAISE_MSG(
>         OE_PLATFORM_ERROR,
> 	"CreateEnclave failed (err=%#x)",
>         enclave_error);
> }
> 
> secs->base = (uint64_t)base;
> ```

@CodeMonkeyLeet - I did not modify the section that you reference. I only modified the 
#elif defined(_WIN32) section with the create_enclave function. This is the section that gets run with the current AESM flow on Windows.

Reply:
We can add the retry if needed to this section once we have support for Windows CI with DCAP and hit this issue. This is the reason that I did not make any changes to this section.